cmd/lassie/fetch.go

```diff
@@ -0,0 +1,186 @@
+package main
+
+import (
+	"context"
+	"fmt"
+	"os"
+	"path/filepath"
+	"strings"
+	"time"
+
+	"github.com/dustin/go-humanize"
+	"github.com/filecoin-project/go-fil-markets/retrievalmarket"
+	"github.com/filecoin-project/lassie/cmd/lassie/internal"
+	"github.com/filecoin-project/lassie/pkg/client"
+	"github.com/filecoin-project/lassie/pkg/eventpublisher"
+	"github.com/filecoin-project/lassie/pkg/indexerlookup"
+	"github.com/filecoin-project/lassie/pkg/retriever"
+	"github.com/google/uuid"
+	blocks "github.com/ipfs/go-block-format"
+	"github.com/ipfs/go-cid"
+	"github.com/ipfs/go-datastore"
+	dss "github.com/ipfs/go-datastore/sync"
+	blockstore "github.com/ipfs/go-ipfs-blockstore"
+	carblockstore "github.com/ipld/go-car/v2/blockstore"
+	"github.com/libp2p/go-libp2p/core/peer"
+	"github.com/multiformats/go-multiaddr"
+	"github.com/urfave/cli/v2"
+)
+
+func Fetch(c *cli.Context) error {
+	if c.Args().Len() == 0 || !c.IsSet("output") {
+		return fmt.Errorf("usage: lassie fetch -o <CAR file> [-t <timeout>] <CID>")
+	}
+	progress := c.Bool("progress")
+
+	rootCid, err := cid.Parse(c.Args().Get(0))
+	if err != nil {
+		return err
+	}
+
+	configDir, err := setupConfigDir()
+	if err != nil {
+		return err
+	}
+
+	carStore, err := carblockstore.OpenReadWrite(c.String("output"), []cid.Cid{rootCid})
+	if err != nil {
+		return err
+	}
+
+	var blockCount int
+	putCb := func(putCount int) {
+		blockCount += putCount
+		if !progress {
+			fmt.Print(strings.Repeat(".", putCount))
+		} else if blockCount%10 == 0 {
+			fmt.Printf("Received %d blocks...\n", blockCount)
+		}
+	}
+	bstore := &putCbBlockstore{parent: carStore, cb: putCb}
+	retriever, err := setupRetriever(configDir, c, c.Duration("timeout"), bstore)
+	if err != nil {
+		return err
+	}
+
+	fmt.Printf("Fetching %s", rootCid)
+	if progress {
+		fmt.Println()
+		retriever.RegisterListener(progressPrinter{})
+	}
+	stats, err := retriever.Retrieve(c.Context, rootCid)
+	if err != nil {
+		fmt.Println()
+		return err
+	}
+	fmt.Printf("\nFetched [%s] from [%s]:\n"+
+		"\tDuration: %s\n"+
+		"\t  Blocks: %d\n"+
+		"\t   Bytes: %s\n",
+		rootCid,
+		stats.StorageProviderId,
+		stats.Duration,
+		blockCount,
+		humanize.IBytes(stats.Size),
+	)
+
+	carStore.Finalize()
+
+	return nil
+}
+
+func setupConfigDir() (string, error) {
```

Comment:
I would say for the CLI, just make a new peer key on each run, unless it matters, to avoid a saved datastore on the users system.

It will make sense for sure for the HTTP daemon when we get to it.